crates/matrix-sdk/src/room/timeline/read_receipts.rs

```diff
@@ -0,0 +1,228 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::HashMap, ops::Deref, sync::Arc};
+
+use eyeball_im::ObservableVector;
+use indexmap::IndexMap;
+use ruma::{
+    events::receipt::{Receipt, ReceiptEventContent, ReceiptThread, ReceiptType},
+    EventId, OwnedEventId, OwnedUserId, UserId,
+};
+use tracing::error;
+
+use super::{
+    inner::{RoomDataProvider, TimelineInnerState},
+    rfind_event_by_id, EventTimelineItem, TimelineItem,
+};
+
+struct FullReceipt<'a> {
+    event_id: &'a EventId,
+    user_id: &'a UserId,
+    receipt_type: ReceiptType,
+    receipt: &'a Receipt,
+}
+
+pub(super) fn handle_explicit_read_receipts(
+    receipt_event_content: ReceiptEventContent,
+    own_user_id: &UserId,
+    timeline_state: &mut TimelineInnerState,
+) {
+    for (event_id, receipt_types) in receipt_event_content.0 {
+        for (receipt_type, receipts) in receipt_types {
+            // We only care about read receipts here.
+            if !matches!(receipt_type, ReceiptType::Read | ReceiptType::ReadPrivate) {
+                continue;
+            }
+
+            for (user_id, receipt) in receipts {
+                let ReceiptThread::Unthreaded = receipt.thread else {
+                    continue;
+                };
+
+                let receipt_item_pos =
+                    rfind_event_by_id(&timeline_state.items, &event_id).map(|(pos, _)| pos);
+                let is_own_user_id = user_id == own_user_id;
+                let full_receipt = FullReceipt {
+                    event_id: &event_id,
+                    user_id: &user_id,
+                    receipt_type: receipt_type.clone(),
+                    receipt: &receipt,
+                };
+
+                if maybe_update_read_receipt(
+                    full_receipt,
+                    receipt_item_pos,
+                    is_own_user_id,
+                    &mut timeline_state.items,
+                    &mut timeline_state.users_read_receipts,
+                ) && !is_own_user_id
+                {
+                    // Update the new item pointed to by the user's read receipt.
+                    if let Some((pos, mut remote_event_item)) = receipt_item_pos.and_then(|pos| {
+                        timeline_state.items[pos]
+                            .as_event()
+                            .and_then(|e| e.as_remote())
+                            .map(|e| (pos, e.clone()))
```

Comment:
```suggestion
                        let e = timeline_state.items[pos]
                            .as_event()?
                            .as_remote()?;
                        Some((pos, e.clone())
```

Reply:
Also moving the expression after the `=` of `if let` into a binding might make things more readable again.